java/ql/src/Diagnostics/CoverageDiagnostics.ql

```diff
@@ -0,0 +1,18 @@
+/**
+ * @name Diagnostics for framework coverage
+ * @description Expose diagnostics for the number of API endpoints covered by CSV models.
+ * @kind diagnostic
+ * @id java/diagnostics/framework-coverage
+ */
+
+import java
+import semmle.code.java.dataflow.ExternalFlow
+
+from string packageAndType, int rows
+where
+  exists(string package, string type |
+    modelCoverage(package, _, _, type, _) and
+    packageAndType = package + ";" + type and
+    rows = sum(int n | modelCoverage(package, _, _, type, n) | n)
```

Comment:
This sum is wrong. You're adding all the _different_ ints that satisfy `modelCoverage(package, _, _, type, n)`, so if two kinds happen to have the exact same number of models, then those numbers won't be added. You'll need to add `string kind` in the range. Another minor thing: If you change `sum` to `strictsum`, then it'll bind `package` and `type` so you can drop the first line.
```suggestion
    packageAndType = package + ";" + type and
    rows = strictsum(int n, string kind | modelCoverage(package, _, kind, type, n) | n)
```

Reply:
Thanks for the explanation. The `kind` thing is tricky, that's hard to see. 

Regarding `strictsum` - do you think it might be worth mentioning that on https://codeql.github.com/docs/ql-language-reference/expressions/ ?
Currently it just says `strictconcat, strictcount, and strictsum: These aggregates work like concat, count, and sum respectively, except that they are strict. That is, if there are no possible assignments to the aggregation variables that satisfy the formula, then the entire aggregation fails and evaluates to the empty set (instead of defaulting to 0 or the empty string). This is useful if you’re only interested in results where the aggregation body is non-trivial.`.

Reply:
No, the missing `kind` in the range has nothing to do with the switch from `sum` to `strictsum`. The explanation of `strictsum` that you linked seems perfectly adequate.